### PR TITLE
JS: Make type information more accessible

### PR DESCRIPTION
**Pull request by asger-semmle**
This PR takes a few steps towards making type information more accessible:
1. Add `Node.hasUnderlyingType`. This takes care of the mapping between AST and DataFlow graph and handles TypeScript types and JSDoc types.
2. Make the TypeScript extractor associate file names with module names in more cases (so the above predicate works in more cases).
3. Associate module names with package.json names within the same project, so the above works for packages that are part of the snapshot.
4. Use the above features in the Firebase model. More will follow, jQuery in particular, but this is meant as a "first adopter" to avoid a huge PR.

The last three commits deal with some perf issues that were triggered by this. One of the problems that came up is that calling `getFile()` on some type `T` would cause location info to be computed for all locatables, not just `T`, due to `getFile()` dispatching to `Locatable.getFile` which depends on `Locatable.getLocation`.

Evaluations: [full eval on smoke-test](https://git.semmle.com/asger/dist-compare-reports/tree/more-type-info_1567089016648) and [security on typescript-full](https://git.semmle.com/asger/dist-compare-reports/tree/more-type-info_1567091310247).

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/TypeScript.qll [hunk 1]
```diff
@@ -1550,14 +1550,20 @@ class ReferenceImport extends LineComment {
   string getAttributeName() { result = attribute }
 
   /**
+   * DEPRECATED. This is no longer supported.
```
Comment (xiemaisi) [inline javascript/ql/src/semmle/javascript/TypeScript.qll]: OOI, why do we need to drop this?

Comment (asger-semmle) [reply javascript/ql/src/semmle/javascript/TypeScript.qll]: It caused location info to be computed for all line comments.

Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/TypeScript.qll]: Ugh, yeah, I can see that being a problem.

**Comment by asger-semmle** [comment]
> Do we want a change note for the Firebase improvements?

Sure. There's already 2+ PRs open that create the change note file, so I'll wait for one of those to land.

**Comment by asger-semmle** [comment]
Added the change note.

**Pull request merged**